### ISSUE TITLE
Jenkins gets stuck when there are no resources to schedule a container

### ISSUE THREAD

**Issue by gamma** [jenkinsci/docker-plugin#594]
If you get some troubles with docker-plugin, please report
 - [ x ] docker-plugin version you use: **1.1+**
 - [ x ] jenkins version you use: **2.96**
 - [ x ] docker engine version you use: **Version = swarm/1.2.6, API Version = 1.22**
 - [ ] stack trace / logs / any technical details that could help diagnose this issue

I have my docker cloud images configured to only spawn 5 to 10 instances max. There are several jobs that start 5+ jobs at once. The first jobs get started until the limit is reached and Jenkins would wait for resources being available.

Starting with docker-plugin 1.1 this does not work as expected. The jobs get stuck. Even stopping and starting new jobs manually does not work, only a restart of Jenkins.

**Comment by MatthewChiappa** [comment]
I'm experiencing the same issue with with the same setup only with Jenkins at version 2.73. Every day I can slowly see the number of jobs stuck in the queue. The only way to get the jobs to run is to kick off more builds and it gets forced out of the queue. The only cure to this is to restart Jenkins, but after a day or so the problem keeps happening. We also run 20+ jobs at a time at some points, which hits our container cap.

**Comment by ndeloof** [comment]
I still don't fully understand the NodeProvisioned logic and how to get eventually rid of such issues.

**Comment by pjdarton** [comment]
@ndeloof My understanding is that Jenkins' handling of all this is rather basic (possibly _too_ basic).  All responsibility for making it work falls upon Cloud's [provision(Label label, int excessWorkload)](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L234) method.

This method will be called, repeatedly, as long as Jenkins has jobs in the queue. e.g. if Jenkins has a job queued that needs the label "unix" then provision will be called and expected to return a PlannedNode that'll (eventually) give a Node that's got the "unix" label that can be used to run the job.  Actual provisioning runs in [another thread](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L263) but eventually gives Jenkins a new Node, that'll then run one or more jobs and reduce demand.

Side note: There's a shortcoming with this API - even the plugin has already returned a PlannedNode that'll eventually create a Node then, until that Node finally exists, Jenkins will keep on calling the provision method.  It's up to the provision method to keep a note of all the executors that are in the process of being created so it doesn't over-provision by the time all the PlannedNodes become real Nodes.  This is less of a problem with Docker as container provisioning is fast, but e.g. the vSphere and OpenStack plugins have to cope with a lengthy delay between the request for a new VM and the VM coming online.  I suspect that this issue is because a PlannedNode doesn't state its labels, just the number of executors, so Jenkins can't tell what it's going to get.  My guess is that this issue is why the [provisionedImages](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L97) field was added.

However, in this case, my *guess* is that the most likely reason for the plugin failing to create new containers would be a (mistaken) belief that it's already running too many of them.  It looks like the [provisionedImages](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L97) field is static and only considers the image name as an index.  That'll mean that, in the scenario where one has multiple docker clouds defined, but which share the same image (e.g. a single image that can be run in multiple places), the plugin will confuse containers being started on one docker host for containers on any other.
Maybe I'm mis-reading the code here, but it looks like [this code](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L467) will be adding the count of _all_ "in progress" containers to the count of containers for a single cloud, i.e. if one docker host is slow and has 10 containers in the process of starting, it'll mean that the plugin thinks that all the other docker hosts are running 10 containers more than they really are.
I'd expect this sort of thing to show up in the logs though, as the plugin logs its reasons for not provisioning [here](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L473) and [here](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L478).

Also, [decrementAmiSlaveProvision](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L221) does not exactly undo the effects that [addProvisionedSlave](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L453) has on the provisionedImages field (because of the exception handling in decrementAmiSlaveProvision), so this could result in the field containing bad data if things get called in an unexpected order.  It'd be non-trivial to work out if this could ever happen in practice (multi-threaded & multi-instance stuff gets tricky to understand), but it wouldn't be difficult to refactor the code to ensure it can't ever be a problem.

TL;DR: Beware of static fields and beware of increment and decrement methods not being perfect opposites.

**Comment by ndeloof** [comment]
Your description of  `provision` method is right, I just haven't fully understand what makes this method to be invoked from NodeProvisioner, and how to efficiently reduce this delay. But AFAIK it handle planned nodes so it won't call provision again as a node is pending, so failure to provision has to be reported using `plannedNode.completeExceptionally()`.

About the template instance counter, it is indeed based on image ID, not considering other template parameters or owning cloud. A better approach would be to rely on labels to flag container for a specific template, and count matching containers with a filter.

> if one docker host is slow and has 10 containers in the process of starting, it'll mean that the plugin thinks that all the other docker hosts are running 10 containers more than they really are.

The `amiCount` logic (notice the variable name: so this code is a copy paste from EC2 plugin, what a good news :-\) do indeed count containers for a cloud, which is associated to a docker host (or swarm cluster). So I don't really get your comment: another cloud to configure access to a distinct docker host will have a dedicated counter.

I assume the root cause for this issue is some race condition in `provisionedImages` increment/decrement logic. If you can't see repeated `Not Provisioning ...` in logs, then the provision() method is not even invoked by NodeProvisioner, which still believe a PlannedNode is finishing it's setup. If you can see it, it will log the number of container per cloud / template and help diagnose a counter leak.

Anyway, I think this whole logic should be deprecated and replaced by a resource counter, based on available resources on dockerhost _vs_ template memory and/or cpu constraints. We could just let docker host reject container creation using it's own internal resource management logic. As this is already implemented in docker daemon, I don't see it useful to re-implement such resource management in Jenkins (and have to maintain/fix it).

**Comment by pjdarton** [comment]
In my experience, the provision method gets called by Jenkins
1. When Jenkins first gets a queued job that it can't run on an existing agent/slave/node.
2. Periodically while there are queued jobs

If the plugin decides not to create a new container when it is first called, it might be 60 seconds or so before it gets called again ... and if you know of a way of reducing that delay, both the OpenStack plugin and VSphere plugin would benefit too ;-)

In my experience (using the docker plugin) the main cause of the docker plugin failing to start new containers is because the docker daemon has multiple containers running that are not successfully-connected Jenkins slaves - it's quite easy to end up with "orphan" containers that are running fine as far as docker is concerned but which are not known to Jenkins.
(FYI I implemented a non-trivial amount of orphan-collection logic on the VSphere plugin because this suffered from this issue.  There's similar anti-orphan functionality in the OpenStack plugin too.)

Just because code was copied doesn't guarantee that it is correct - sometimes it's only correct in the first place because of other factors not present in the second place (and sometimes it's wrong in two places).
My issue with the amiCount logic is that provisionedImages is **static**, so this field will contain a count of all the containers being provisioned for ALL docker clouds (i.e. all instances of DockerCloud), not just the cloud that [this DockerTemplate](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L453) belongs to.
What's needed is to maintain a count of "containers currently being started on the docker daemon pointed to by our DockerCloud's docker API endpoint", and to keep that count on a per-template basis.  That means that if provisionedImages remains static, it'll have to be a Map<Cloud, Map<Image, Integer>> rather than just a Map<Image, Integer>.  Either that or the "key" value used to access provisionedImages should be a composite of both the cloud API endpoint and the template.

I'm also a little concerned that there's no guarantee that decrementAmiSlaveProvision will get called after each successful call to addProvisionedSlave - I'd like to see a try ... catch ... finally block immediately after the addProvisionedSlave to ensure that, unless the code successfully queues the Runnable [here](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L263), decrementAmiSlaveProvision will get called.
I know that there isn't much to go wrong in between those two points in the code, but unless we add a means of re-calculating the correct contents of provisionedImages when errors happen, we need to be 100% sure that it can never become incorrect, even when exceptions are getting thrown.

...also, if decrementAmiSlaveProvision gets called when there's no slave provision in progress, we should log an error rather than just [suppressing it](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L227).  While this line _should_ never get called, if we've got a mistake somewhere and the threading doesn't behave the way we expect then it'd be better to know this than to remain ignorant.

Re: root cause / logs
FYI the reason I've got involved here is because one of my work colleagues noticed this kind of problem on one of our Jenkins servers.  I've asked than "next time" they capture the logs so hopefully we'll get some better information if it happens again.
i.e. We're going to keep a watchful eye on the logs so that, hopefully, we'll be able to give some definite information instead of generalities.

Re: replacing logic:  Yes, I agree that, ideally, we'd have some nice configuration options to let us set up some load-balancing options or, ideally, tell Jenkins to automatically load-balance and have it interrogate docker to do that without much human decision making required (FYI I implemented similar functionality on the VSphere plugin) but I don't think it'd be trivial to do that in a way that won't cause other performance issues (it's important that the provision method completes quickly).

**Comment by ndeloof** [comment]
> Just because code was copied doesn't guarantee that it is correct 

I was kidding on this topic, being a copy/paste from EC2 plugin is not something I consider a good news. Also, I wonder this never have been included in jenkins-core Cloud API so various cloud-provisioner plugin won't have to address this common requirement.

I didn't noticed `amiCount` was static, that's a terrible design issue :-\

call to `decrementAmiSlaveProvision` need to be fixed as well, as you described.

> Re: replacing logic: Yes, I agree that, ideally, we'd have some nice configuration options to let us set up some load-balancing options or, ideally, tell Jenkins to automatically load-balance and have it interrogate docker to do that without much human decision making required (FYI I implemented similar functionality on the VSphere plugin) but I don't think it'd be trivial to do that in a way that won't cause other performance issues (it's important that the provision method completes quickly).

Actually, if we enforce *all* template do define a memory constraint (maybe by having a reasonable default value at cloud configuration level) then the docker daemon will just reject node provisioning as we try to create a new agent, and will automatically retry until some resource become available. So we wouldn't have to implement anything on master.

**Comment by pjdarton** [comment]
I _suspect_ that the reason that all plugins have to implement this code is because the core-code can't, because the PlannedNode doesn't give enough information for it to do so, and changing a core API is hard work because of all the compatability issues it causes.  I'm sure it'd be possible, but it wouldn't be easy and it wouldn't be quick.

I don't think using a static is necessarily a bad design decision - we need this field to survive someone reconfiguring the cloud(s) via Manage Jenkins -> Configure (which can create/delete DockerCloud instances) while the plugin is busy working - but the indexing of the Map must take the shared nature into consideration.
We just have to be mindful that the synchronization locking on this field will block all DockerClouds, not just one instance, so we need to avoid doing any I/O operations while this is locked.

Re: docker daemon will just reject node provisioning
If we do this then we will then have to implement some functionality on the master so that it takes previous node-provisioning rejections into consideration.
If we don't do something like that, then the first cloud will get full, start rejecting requests, but we'll never ask the other clouds to start container.
e.g. Lets say we've got one job in the queue, two equally-capable DockerClouds, and our first DockerCloud is full enough that it will reject the next node provisioning request.  The plugin will (currently) see that this DockerCloud has a suitable Template, queue a request for a new container, and then the provision method will return that PlannedNode.  It will ignore the fact that our second DockerCloud could have provided a node because it simply chose the first available template.  Some time later, the docker daemon of the first DockerCloud will reject the request (because it's full), so Jenkins doesn't get the Node it wanted.  So, after a minute, Jenkins asks the plugin to provision again ... but unless we change our template ordering, the plugin code would choose the first DockerCloud again.

We would need to keep some form of records of our current/previous requests, for each template, of when requests for containers were made and when they completed or failed, so that we could then take that information into consideration when deciding which template to try.  e.g. sort templates by last-failure-time (with last-request time as a tie-breaker) so that we only try templates that rejected our last request once we've tried all the others.
At least, that was the conclusion I came to when I wrote [this](https://github.com/jenkinsci/vsphere-cloud-plugin/blob/master/src/main/java/org/jenkinsci/plugins/vsphere/tools/CloudProvisioningAlgorithm.java) for the VSphere plugin.

**Comment by ndeloof** [comment]
What I mean is jenkins-core do already provide `AbstractCloudImpl` with `instanceCap` attribute, but not the logic for `provision()` to support this upper bound limit. So every plugin has to re-implement, and such code duplication could be avoided if some base classes to cover this use case where provided (for sample with some AbstractCloudSlaveTemplate / AbstractPlannedNode ...)

As you said, Cloud instance being replaceable, we miss some place to store the live cloud state. Can either re-compute on each attempt to provision (doesn't seem to me this would be such a pain using docker ps with adequate filter), or need _some way_ to store this, which Cloud API doesn't provide. 

> If we don't do something like that, then the first cloud will get full, start rejecting requests, but we'll never ask the other clouds to start container.
IIUC you have multiple Clouds configured with same templates and labels, just to target multiple docker physical hosts. Can't you use docker swarm standalone ? We could also have DockerCloud support multiple hosts URIs, not just one, and do some basic load balancing

**Comment by pjdarton** [comment]
Ah, I've not investigated AbstractCloudImpl before.  My experience in this area has been to assist with a PR on the docker plugin (years ago), some non-trivial enhancements to the OpenStack cloud plugin (last year) and some serious re-working/enhancements/bugfixing of the VSphere cloud plugin (during the last 2-3 years), but in each case I wasn't looking to redesign the architecture of the plugin so didn't pay much attention to what the Jenkins core provided (aside from whatever was already in use by the plugin I was working on).
I think that what we (as in "all Jenkins users") really need is someone with a lot of Jenkins-core knowledge to take a good look at all the current "cloud" plugins, identify common functionality and patterns, and then implement that in the core code so that all plugins can benefit from "best practice".  However that's quite a lot of work.

Re: re-compute
We want to avoid having to do docker API calls during the thread that's calling provision(...).  This method call needs to complete fairly swiftly and be very reliable, so that means we have to avoid making docker API calls on a routine basis.
What I've seen "in the real world" is that, sometimes, the docker daemon crashes/locks-up such that it'll accept API calls, but never complete them.  This results in any docker-API call hanging indefinitely, so if that happens during provision(...) then it'll prevent *all* clouds from working.  I've found that this is usually followed by angry developers sitting by my desk asking "when is Jenkins going to be working again" :-)
I've also seen similar issues with VSphere and OpenStack - basically, any remote API call should be done outside the main provision(...) call (that's why we run the code that'll get the new container started in a separate thread) - if we're going to self-police on instance counts, we should do so by keeping a count internally (much like the existing code tries to do, albeit poorly at present).

Yes, we have multiple clouds configured with the same templates and labels just to target multiple docker hosts.
Yes, we could use docker swarm, but we started using docker before this was an option (before basic docker was stable!) and so my total knowledge about docker swarm is that it sounds like it fits our needs perfectly - other work pressures have meant that we've not spent any time investigating this.
i.e. it remains on our "TODO" list (although we might jump straight to using Kubernetes).

However, although we could probably benefit from docker swarm, the use case of having the same image on different hosts is perfectly valid - different hosts might launch it with different options, and give slightly different Jenkins-slave capabilities as a result.  The plugin should cope with that.
i.e. just because _my_ Jenkins DockerCloud configuration looks like a copy/paste accident doesn't mean everyone's will ;-)

The possibility of having DockerCloud support multiple host URIs had occurred to us as well, and we (briefly, a few years ago now) considered adding that enhancement before other work pressures took our attentions elsewhere.  These days, I do not think that it's something the plugin should do because anyone wanting this kind of thing should be using a docker swarm or similar - just as you stated earlier, load balancing should be the API-endpoint's job.

**Comment by ndeloof** [comment]
My suggestion to query docker infra for current number of allocated nodes per template was implicitly suggesting a "with timeout and cached result" to ensure `provision` can be efficiently implemented.

**Comment by pjdarton** [comment]
Good news!  We've caught Jenkins "in the act" of failing to start containers.
It happened twice over the course of the weekend - once on Sunday, and again on the Monday morning shortly after we thought rebooting Jenkins had fixed it.

The bad news is that this is not going to be simple to fix...

First thing to clarify is that that my earlier suspicion of our misuse of the ```provisionedImages``` field seems to be unfounded.  I guess that, while the misuse of the ```provisionedImages``` field isn't doing us any favors, it isn't hurting too much either.  i.e. provisionedImages does _not_ appear to be the underlying cause of this problem.

However, what I *did* see was was lots of ```Computer.threadPoolForRemoting``` threads all which were all ```WAITING on java.util.concurrent.locks.ReentrantLock$NonfairSync@50917a54 owned by "Computer.threadPoolForRemoting [#97]" Id=401```.
The stack of the thread owning that lock was:
```
"Computer.threadPoolForRemoting [#97]" Id=401 Group=main WAITING on java.util.concurrent.CountDownLatch$Sync@34c3cdbd
	at sun.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.CountDownLatch$Sync@34c3cdbd
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at com.github.dockerjava.core.async.ResultCallbackTemplate.awaitCompletion(ResultCallbackTemplate.java:92)
	at com.github.dockerjava.netty.InvocationBuilder$ResponseCallback.awaitResult(InvocationBuilder.java:60)
	at com.github.dockerjava.netty.InvocationBuilder.get(InvocationBuilder.java:189)
	at com.github.dockerjava.netty.exec.ListContainersCmdExec.execute(ListContainersCmdExec.java:44)
	at com.github.dockerjava.netty.exec.ListContainersCmdExec.execute(ListContainersCmdExec.java:15)
	at com.github.dockerjava.netty.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21)
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35)
	at com.nirima.jenkins.plugins.docker.DockerCloud.countCurrentDockerSlaves(DockerCloud.java:437)
	at com.nirima.jenkins.plugins.docker.DockerCloud.addProvisionedSlave(DockerCloud.java:460)
	-  locked com.nirima.jenkins.plugins.docker.DockerCloud@57b79e32
	at com.nirima.jenkins.plugins.docker.DockerCloud.provision(DockerCloud.java:249)
	-  locked com.nirima.jenkins.plugins.docker.DockerCloud@57b79e32
	at io.jenkins.docker.FastNodeProvisionerStrategy.applyFoCloud(FastNodeProvisionerStrategy.java:72)
	at io.jenkins.docker.FastNodeProvisionerStrategy.apply(FastNodeProvisionerStrategy.java:44)
	at hudson.slaves.NodeProvisioner.update(NodeProvisioner.java:320)
	at hudson.slaves.NodeProvisioner.access$000(NodeProvisioner.java:61)
	at hudson.slaves.NodeProvisioner$1.run(NodeProvisioner.java:176)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)

	Number of locked synchronizers = 2
	- java.util.concurrent.ThreadPoolExecutor$Worker@52535615
	- java.util.concurrent.locks.ReentrantLock$NonfairSync@50917a54
```
What seems to have happened is that a call to ```countCurrentDockerSlaves``` aimed at one of my docker hosts hung, and this then deadlocked all of the cloud functionality in Jenkins in a "cascade of doom" - if the ```provision``` method doesn't return, Jenkins can't ask any other cloud to do anything.
Unfortunately, when _any_ cloud plugin's ```provision``` method hands, it takes out _all_ cloud functionality in Jenkins, so we need to ensure that this can't happen.
(I guess another thing would be to enhance the Jenkins core so that it protects itself against this -  that wouldn't fix this bug, but it would reduce the impact of it on other clouds)

So, what we're going to need is
1) Some kind of "timeout" around every docker-API call, so that we avoid permanent deadlock and can also detect a failed/uncooperative docker host.
2) To avoid trying to provision on a failed/uncooperative host.

I guess that, as a first priority, we only really need to ensure that the code called from ```provision``` is 100% reliable - everything else will be in its own thread and so, if that deadlocks, it'll have less of an impact on Jenkins' overall functionality.  However, if adding timeouts around docker-API calls during the ```provision``` method seems to fix this issue then we should really extend that robustness to the rest of the plugin - in my experience, any docker API call _can_ hang, so we need to ensure that its failing don't cause us to fail.

I would also suggest that the ```countCurrentDockerSlaves``` method also be enhanced so that it caches the [list of containers](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L434) (for a short period) in order to minimize the number of remote API calls - at present, we're calling this fairly-expensive API repeatedly.  Obviously, as we're going to affect the count if we successfully provision a container then we'll need to be careful here.

However, the top priority I think is to ensure that ```countCurrentDockerSlaves``` _cannot_ deadlock, no matter what the remove docker server does.  We need this method to either complete within seconds or explode with an exception.
(maybe "within seconds" should be configurable, but initially I would suggest a hard-coded 60 second timeout would be acceptable, or perhaps we can use the "Connection Timeout")

**Comment by ndeloof** [comment]
good catch.
this synchronous implementation to check infra for current load is indeed terrible, I don't know another Cloud implementation we could get inspiration from for a better design.

**Comment by pjdarton** [comment]
The VSphere plugin has a fairly robust way of dealing with this, but I wrote a lot of that so I'm rather biased (it might not necessarily be a good/recommended way of doing things, and it might not be to everyone's tastes, but I'm pretty pleased with it).  In the VSphere code, I (mostly) abandoned any attempt at counting the total number of VMs in the cloud (I found that VSphere only ever returned zero when it was asked - I guess it's very dependent on how VSphere was deployed, but my VSphere server always answered zero) and counted the ones that the plugin was responsible for.  As it was only counting the ones it was responsible for, it could keep an in-memory map, so no need to talk to the remote server except to start and stop slaves.
So, we could copy VSphere's in-memory model of what slaves are being-created, active-and-wanted and active-but-unwanted (i.e. pending deletion) - that allows easy policing of per-template instance counts - but that wouldn't let us set a total container count that included both all the containers we'd started and also any containers that _other_ things (e.g. other Jenkins servers) had started.
Note: because it'd be impossible to prevent two independent Jenkins servers (that are both accessing the same docker host) from simultaneously seeing "there's only 9 containers running and my instance cap is 10" and thus both deciding to start a new container, bringing the total to 11 (1 more than the cap), I don't think we're under any obligation to prevent that scenario.  So, what we could do instead of to keep a list of the containers that are _not_ ours in a cached list (that we refresh periodically) so that we can "mostly" avoid overloading a host, while keeping a list of the containers that we are responsible for in an in-memory list that we maintain ourselves.
That'd mean that users could set a total container cap which is likely to prevent excessive over-provisioning (but without causing calls to the docker API during the ```provision``` method), and/or a containers-from-this-Jenkins cap which _will_ set a guaranteed limit, and/or a containers-from-this-template cap (also guaranteed).

While on the subject of instance caps, I would point out that there are 3 kinds of instance cap:
1) The total number of containers the host should be asked to run at any time, from all sources combined.  This allows us to have multiple Jenkins servers pointing at the same Docker host and to (unfairly) share it.
2) The total number of containers that _this_ Jenkins is allowed to ask the host to run at once.  If this is specified and (1) is also set then this must be no greater than (1).
3) The maximum number of containers we're allowed to have running from a single template at any one time.  This should be mandatory if (1) and (2) are not set - we need a limit somewhere, even if it's a large one.  Obviously, this must be no greater than (2) or (1).

At present, the plugin only implements (1) and (3), but if we deprecated (1) in favor of implementing (2) then we'd never need to routinely count the number of running containers in the docker host again, as we could rely on our internal count being correct - that's what I did with the VSphere plugin, and I'm pretty happy with how that turned out.
Note: We would still have to get a list of containers on startup, so that, after a reboot, we could re-discover any containers that we'd asked to be started just before Jenkins crashed, otherwise one can end up with containers in docker that Jenkins doesn't know about.


That all said, I've been digging through the code and I suspect that adding timeouts might not be so tricky after all...
It looks like [DockerAPI](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/io/jenkins/docker/client/DockerAPI.java#L131) could easily be enhanced to call ```withReadTimeout``` as described [here](https://github.com/docker-java/docker-java/wiki#intialize-docker-client-advanced), although that would need us to change [DockerAPI](https://github.com/jenkinsci/docker-plugin/blob/master/src/main/java/io/jenkins/docker/client/DockerAPI.java#L130) to use [JerseyDockerCmdExecFactory](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/jaxrs/JerseyDockerCmdExecFactory.java) instead of ```NettyDockerCmdExecFactory```.

What do you think?  Do you know if there was a good reason for choosing ```NettyDockerCmdExecFactory``` over ```JerseyDockerCmdExecFactory```?  (I'm not familiar with either of them. so I don't know what the pros/cons are)

If we have to use Netty then a read-timeout would still be possible, but we'd have to use an enhanced version of NettyDockerCmdExecFactory where we had extra code
```
    private Integer readTimeout = null;

    /**
     * Configure read timeout in milliseconds
     */
    public NettyDockerCmdExecFactory withReadTimeout(Integer readTimeout) {
        this.readTimeout = readTimeout;
        return this;
    }
```
and an enhanced [configure](https://github.com/docker-java/docker-java/blob/master/src/main/java/com/github/dockerjava/netty/NettyDockerCmdExecFactory.java#L747) method with additional code
```
        if (readTimeout != null) {
            channel.pipeline().addLast("readTimeoutHandler", new ReadTimeoutHandler(readTimeout, TimeUnit.MILLISECONDS));
        }
```
in it, assuming I've read the [Netty docs](https://netty.io/4.0/api/io/netty/handler/timeout/ReadTimeoutHandler.html) correctly.  That would be ugly as ```configure``` is private so we'd need to duplicate a lot of the class code to add those lines, but I'd choose "ugly" over "broken" any day...

**Comment by ndeloof** [comment]
>  counted the ones that the plugin was responsible for

I'm +1 with this. For legacy reasons (sic) docker-plugin was considering all containers running the same image. With docker support for labels we can now track which container has been created for a specific jenkins label, and as such make a better computation.

It sounds to me keeping the counter internal to jenkins master and only asking the docker infra on startup would make perfect sense.

The reason we rely on NettyDockerCmdExecFactory is that Jersey dependency have many collisions with jenkins-core dependencies and other plugins. 
That being said I'd like to give up with docker-java, and either move to [jocker](https://github.com/dockins/jocker) or [fabric8/docker-client](https://github.com/fabric8io/docker-client) as both use a generated API and as such a kept in sync with docker API (docker-java has to be manually updated in a heavy way for any addition to the API, so is missing many options)

**Comment by MatthewChiappa** [comment]
I saw this was recently committed https://github.com/jenkinsci/docker-plugin/commit/5a2a123cd808a9c82d154cbcc22ef906a6a97907 would this fix the issue I described above? @ndeloof

**Comment by pjdarton** [comment]
I've created PR #610 that uses an enhanced ```NettyDockerCmdExecFactory``` to support ```withReadTimeout(milliseconds)``` and then adds a ```readTimeout``` field to DockerCloud/DockerAPI etc.
I've tested it locally and it seems to make the plugin more robust, in that it now times-out if a docker daemon locks up _but_ I've yet to deploy it on my primary Jenkins server for long enough to prove that it's resolved the "one dead docker = totally dead Jenkins" issue that I described [above](https://github.com/jenkinsci/docker-plugin/issues/594#issuecomment-357663827).

FYI I've also raised a [PR on docker-java](https://github.com/docker-java/docker-java/pull/982) to implement ```readTimeout``` in their ```NettyDockerCmdExecFactory```, so if that gets included in the next docker-java release then it'll simplify the code in docker-plugin.

Note: having spent some time delving into Netty + docker-java to do this, I'm not very impressed with either of them, so I've got nothing against ditching docker-java either in the longer term in favor of something else (although I guess I might be equally unimpressed with the alternatives too ;-)

**Comment by crazy-max** [comment]
Still the same problem with version 1.1.3.
I had to downgrade to 1.0.4.

**Comment by pjdarton** [comment]
As of version 1.1.5, we've got some automated cleanup so that, even if Jenkins is unable to remove a dead container at first, it'll keep on trying to remove that container (every 5 minutes by default) until it's gone.
That should result in dead/orphaned containers no longer sitting around and counting against the container limit that you've configured.

So, if folks can upgrade to 1.1.5 and let me know if that's fixed it, that'd be really useful.
If it hasn't fixed it, please grab lots and lots of logs showing when containers are started and when Jenkins decides not to provision any more.

**Comment by crazy-max** [comment]
It seems ok on 1.1.5 release but be careful, the SSH host key verification strategy is resetted while upgrading in the docker template.

**Comment by pjdarton** [comment]
I'm an aware that upgrading from older versions of the docker plugin can result in some data-migration issues.  I'm unsure of the specifics (I wasn't heavily involved in the plugin at the time) but I've seen many anecdotal reports of "bad things happen".  I don't have the time (or, rather, my boss doesn't want me to spend the time!) to fix those issues either.

However, if it's fixed in 1.1.5 then I guess we can close this issue.
